### PR TITLE
feat(operator): add podStatus ui wheel to component

### LINKED ISSUE #874: Support components update on CRD for UI 'ready' wheel
##### What can be improved

Describe a graphical component to display a nice wheel over the status of the session

##### Why

User friendly look :)

```yaml
status:
 readiness:
    components:
      pending: []
      ready:
        - grafana
        - istio-discovery
        - istio-egress
        - istio-ingress
        - kiali
        - mesh-config
        - prometheus
        - telemetry-common
        - tracing
      unready: []
```

### PR DESCRIPTION
**Pull request by aslakknutsen**
fixes #874

Screenshot from 2021-08-17 21-09-52

- [x] Service is only located, not "completed"
- [x] Only displayed via the CRD -> Instance view, not Search -> Session view..

### CODE REVIEW

**Comment by alien-ike** [comment]
Skipping CI for Draft Pull Request.
If you want CI signal for your change, please convert it to an actual PR.
You can still manually trigger a test run with `/test all`

File: api/maistra/v1alpha1/session_types.go [hunk 1]
```diff
@@ -88,6 +88,50 @@ type SessionStatus struct {
...
+	s.Ready = s.removeFrom(s.Ready, comp)
+	s.UnReady = s.removeFrom(s.UnReady, comp)
+	s.Pending = append(s.Pending, comp)
+}
+
+func (s *StatusComponents) SetReady(comp string) {
+	s.Pending = s.removeFrom(s.Pending, comp)
+	s.UnReady = s.removeFrom(s.UnReady, comp)
+	s.Ready = append(s.Ready, comp)
+}
```
Comment (bartoszmajsak) [inline api/maistra/v1alpha1/session_types.go:117]: It looks fine and stuff, but maybe some unit tests with boundary conditions would help from regressions in the future?

File: config/manifests/bases/istio-workspace-operator.clusterserviceversion.yaml [hunk 2]
```diff
@@ -21,6 +21,12 @@ spec:
       displayName: Session
       kind: Session
       name: sessions.maistra.io
+      statusDescriptors:
+      - description: Status of resources deployed/modifed by this Session resource
+        displayName: Resources
+        path: readiness.components
+        x-descriptors:
+        - urn:alm:descriptor:com.tectonic.ui:podStatuses
```
Comment (bartoszmajsak) [inline config/manifests/bases/istio-workspace-operator.clusterserviceversion.yaml:29]: I wonder at which point in time `com.tectonic.ui` will become `io.openshift.ui` and we won't notice ;)

File: api/maistra/v1alpha1/session_types.go (hunk 1 above)
Comment (aslakknutsen) [reply api/maistra/v1alpha1/session_types.go:117]: good point

File: config/manifests/bases/istio-workspace-operator.clusterserviceversion.yaml (hunk 2 above)
Comment (aslakknutsen) [reply config/manifests/bases/istio-workspace-operator.clusterserviceversion.yaml:29]: heh yea.. :)

File: api/maistra/v1alpha1/session_types.go (hunk 1 above)
Comment (aslakknutsen) [reply api/maistra/v1alpha1/session_types.go:117]: Added unit and int level tests https://github.com/maistra/istio-workspace/pull/875/commits/ccdb745a662402b43b249e07ff6a4d37100ca6f7

File: api/maistra/v1alpha1/session_types_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+var _ = Describe("API manipulation", func() {
+
+	Context("when flipping readiness", func() {
+
+		var components v1alpha1.StatusComponents
+		compoentOne := "one"
+		compoentTwo := "two"
+
```
Comment (bartoszmajsak) [inline api/maistra/v1alpha1/session_types_test.go:17]: Nit-picking... but... you have a typo `component` :)

I would normally do suggestion, but it's easier to rename in the IDE :)

File: api/maistra/v1alpha1/session_types_test.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		})
+		It("should set as UnReady", func() {
+			components.SetUnReady(compoentOne)
+
+			Expect(components.UnReady).To(HaveLen(1))
+		})
+
+		It("should flip from pending to ready", func() {
+			components.SetPending(compoentOne)
+			Expect(components.Pending).To(HaveLen(1))
```
Comment (bartoszmajsak) [inline api/maistra/v1alpha1/session_types_test.go:40]: Not sure we really need this kind of assertions. Adds noise perhaps? It's a test fixture after all.

File: controllers/session/session_controller.go [hunk 5]
```diff
@@ -270,6 +264,11 @@ func (r *ReconcileSession) Reconcile(c context.Context, request reconcile.Reques
 						session.Status.Hosts = splitAndUnique(session.Status.Hosts, modified.Prop["hosts"])
 					}
 				}
+				if modified.Success {
+					session.Status.Readiness.Components.SetReady(modified.Kind + "/" + modified.Name)
+				} else {
+					session.Status.Readiness.Components.SetUnReady(modified.Kind + "/" + modified.Name)
```
Comment (bartoszmajsak) [inline controllers/session/session_controller.go:270]: Shouldn't it be `SetUnready`?

File: controllers/session/session_controller_int_test.go [hunk 6]
```diff
@@ -284,6 +284,103 @@ var _ = Describe("Complete session manipulation", func() {
...
+				session := get.Session("test", "test-session1")
+
+				Expect(*session.Status.State).To(Equal(v1alpha1.StateSuccess))
+				Expect(session.Status.Readiness.Components.Pending).To((HaveLen(0)))
+				Expect(session.Status.Readiness.Components.UnReady).To((HaveLen(0)))
+				Expect(session.Status.Readiness.Components.Ready).ToNot((HaveLen(0)))
+			})
+
+			It("should show pending before execution and missing modification, and unready on unsuccess", func() {
+				testLocator := func(ctx model.SessionContext, ref model.Ref, store model.LocatorStatusStore, report model.LocatorStatusReporter) error {
```
Comment (bartoszmajsak) [inline controllers/session/session_controller_int_test.go:310]: Maybe we should move these `func`s outside of the test itself? Might improve readability in my opinion.

File: controllers/session/session_controller_int_test.go [hunk 7]
```diff
@@ -284,6 +284,103 @@ var _ = Describe("Complete session manipulation", func() {
...
+					return &corev1.Pod{}, func(context model.SessionContext, ref model.Ref, store model.LocatorStatusStore, reporter model.ModificatorStatusReporter) {
+
+						// Then - Verify all are in pending stage before modification stage
+						session := get.Session("test", "test-session1")
+						Expect(session.Status.Readiness.Components.Pending).To((HaveLen(3)))
+
+						// When - One is reported success
+						reporter(model.ModificatorStatus{
+							LocatorStatus: store("Pod")[0],
+							Success:       false,
```
Comment (bartoszmajsak) [inline controllers/session/session_controller_int_test.go:349]: `// When - One is reported success` but `Success: false`? Not sure I understand this.

Comment (bartoszmajsak) [reply controllers/session/session_controller_int_test.go:349]: Same reversed logic vs comment in the next step.

Comment (bartoszmajsak) [reply controllers/session/session_controller_int_test.go:349]: I guess it's passing because of weak assertions checking size of the slice vs actual content. Or am I missing sth here?

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]

File: api/maistra/v1alpha1/session_types_test.go (hunk 4 above)
Comment (aslakknutsen) [reply api/maistra/v1alpha1/session_types_test.go:40]: What do you mean this kind of assertion? What's a test fixture?

File: controllers/session/session_controller.go (hunk 5 above)
Comment (aslakknutsen) [reply controllers/session/session_controller.go:270]: it could be...

File: controllers/session/session_controller_int_test.go (hunk 6 above)
Comment (aslakknutsen) [reply controllers/session/session_controller_int_test.go:310]: As long as it's only used here and it contains the // comments which are related to the setup of the test.. then I don't see how jumping around for assertions and logic helps readability.

File: controllers/session/session_controller_int_test.go (hunk 7 above)
Comment (aslakknutsen) [reply controllers/session/session_controller_int_test.go:349]: Just comments that are in the wrong place. But yeah, it's checking sizes of things.

Comment (aslakknutsen) [reply controllers/session/session_controller_int_test.go]: fixed https://github.com/maistra/istio-workspace/pull/875/commits/269a7b34b1ac42c4a21d809f820880405e27cce5

File: api/maistra/v1alpha1/session_types_test.go (hunk 3 above)
Comment (aslakknutsen) [reply api/maistra/v1alpha1/session_types_test.go]: fixed https://github.com/maistra/istio-workspace/pull/875/commits/ecfa131c39f898fb898c5782eae312a1f080bc75

File: api/maistra/v1alpha1/session_types_test.go (hunk 4 above)
Comment (bartoszmajsak) [reply api/maistra/v1alpha1/session_types_test.go:40]: Test fixture - initial state of things on which you perform some logic and validate the outcome. In this case, you have `componentOne` in pending state. I don't think that should be validate. It's anyway tested earlier in dedicated test.

File: controllers/session/session_controller.go (hunk 5 above)
Comment (bartoszmajsak) [reply controllers/session/session_controller.go:270]: I mean `unready` is a word :)

File: controllers/session/session_controller_int_test.go (hunk 6 above)
Comment (bartoszmajsak) [reply controllers/session/session_controller_int_test.go:310]: Maybe it's just me and the way how I see the code. 1:1, nothing can be done here :)

File: controllers/session/session_controller.go (hunk 5 above)
Comment (aslakknutsen) [reply controllers/session/session_controller.go]: fixed https://github.com/maistra/istio-workspace/pull/875/commits/09e0326e8adbf938d417565697d7161f702ee273

File: api/maistra/v1alpha1/session_types_test.go (hunk 4 above)
Comment (aslakknutsen) [reply api/maistra/v1alpha1/session_types_test.go:40]: Ah, yea, that's true.

Comment (aslakknutsen) [reply api/maistra/v1alpha1/session_types_test.go]: fixed https://github.com/maistra/istio-workspace/pull/875/commits/06b6ac92b8d0d065c6477ae263d5b91390708479

**Review by bartoszmajsak — APPROVED** [review]
